### PR TITLE
feat(api): Create import endpoint

### PR DESCRIPTION
**Pull request by pchila**
## This PR

- adds a new import endpoint in api-service for importing configuration and resources by uploading a zip archive 
- add an emptyDir volume as import scratch space to the api-service in the helm charts and service.yaml
- added a couple of env vars for configuration of control plane endpoint and max size of an uncompressed import package in the helm chart and service.yaml
- endpoint is marked as `WIP/beta` since the implementation is still ongoing, will be cleared once the implementation is done

This first implementation provides a simple POST /import handler along with some zip archive handling and sanity checks.

### Follow-up Tasks

Business logic processing the upload will be developed in follow-up PRs (this PR will handle the upload and local storage of the archive, validation and then throws it away)

### How to test

Code has been covered as much as possible by unit tests, however it is possible to test the upload using the swagger UI at /api/swagger-ui/#/Import/import  and providing a project name and a .zip file

### CODE REVIEW

**Review by christian-kreuzberger-dtx — DISMISSED** [review]
Looks good from my point of view. 
Please wait before we merge this.
@thisthat let me know if we want an additional review from people that recently contributed to the api-service itself.

**Review by warber — COMMENTED** [review]
Since there is already a `handlers` folder and because I think there is no reason to put this code inside `pkg`, can you move the new `import` folder into `handlers` ? Also naming the package after a language keyword and having the need to compensate with `_` feels a bit strange. Maybe there is another suitable name for it, like `importer` ?

File: api/pkg/import/zip_package.go
Comment (thisthat) [inline api/pkg/import/zip_package.go:16]:
```suggestion
const ErrorUncompressedSizeTooBig = errors.New("uncompressed size of package exceeds configured maximum size")
```

File: api/pkg/import/zip_package.go
Comment (thisthat) [inline api/pkg/import/zip_package.go:124]: Since we don't do anything fancy with the returned error, I suggest to remove the extra wrapping of a function

File: api/pkg/import/zip_package_test.go
Comment (thisthat) [inline api/pkg/import/zip_package_test.go:117]: I would add also a small test to check that we fail if an invalid zip is supplied

File: api/swagger.yaml
Comment (thisthat) [inline api/swagger.yaml:67]:
```suggestion
      summary: "Alpha: Import a zip package"
      description: "This API is still in alpha state and we do not recommend its usage outside of testing purposes"
```

**Review by thisthat — CHANGES_REQUESTED** [review]

File: api/pkg/import/handler.go
Comment (pchila) [reply api/pkg/import/handler.go:68]: @christian-kreuzberger-dtx any thoughts ?

File: api/pkg/import/handler.go
Comment (pchila) [reply api/pkg/import/handler.go:75]: The asterisk is there to identify the unique part of the filename so that we obtain a file with a name like `importConfig<random part>.zip`:  https://pkg.go.dev/io/ioutil#TempFile

File: api/swagger.yaml
Comment (pchila) [reply api/swagger.yaml:67]: @christian-kreuzberger-dtx please check if this is acceptable

**Comment by pchila** [comment]
@warber regarding the package, I felt it was a little too much code to put all in the handlers package (aside from the fact that the whole zip archive handling does not really fit in there).
Agreed on the poor naming, I would like to hear from @christian-kreuzberger-dtx  if something like `importer` or something equivalent is ok...
Regardless of the final name choice, I would still create a dedicated package for the zip handling and the (upcoming) business logic for the manifest processing if it's ok with you...

File: api/pkg/import/project_checker.go
Comment (pchila) [reply api/pkg/import/project_checker.go:10]: if I understand correctly, you are proposing to use [this](https://github.com/keptn/go-utils/blob/d5ad06c725cddaadb72d60cf0566421fad1f2c12/pkg/api/utils/projectUtils.go#L106) ?

File: api/pkg/import/zip_package.go
Comment (pchila) [reply api/pkg/import/zip_package.go:64]: it was a hint for the next tasks but sure it's a comment and can be removed

File: api/pkg/import/zip_package.go
Comment (pchila) [reply api/pkg/import/zip_package.go:16]: this cannot be done as the right-hand-side expression is not a constant expression.
We have to keep the constant as a global variable in this case

File: api/pkg/import/zip_package.go
Comment (pchila) [reply api/pkg/import/zip_package.go:124]: In this case it's necessary to trigger the defer() calls at the end of each iteration rather than at the end of the function.
The idea is to close  the files as soon as possible (each iteration of the for loop extracts a file) hence the function.
If it looks messy this can be moved to a named private function...

File: api/pkg/import/zip_package_test.go
Comment (pchila) [reply api/pkg/import/zip_package_test.go:117]: There was a corresponding test for the handler before the refactoring, I will remove that and create a new one for the zip archive

File: api/pkg/import/project_checker.go
Comment (thisthat) [reply api/pkg/import/project_checker.go:10]: Exactly :)

**Comment by christian-kreuzberger-dtx** [comment]
@warber We wanted to introduce the `pkg` folder to follow best practices seen in other Keptn integrations and microservices. This way we don't overload the `handler` package. Also, if we ever need to move the implementation, it's way easier if it lives in its own package.

File: api/pkg/import/handler.go
Comment (christian-kreuzberger-dtx) [reply api/pkg/import/handler.go:68]: @thisthat I've told @pchila to avoid that for internal API calls, as in the end it would mean having *multiple* HTTP Status 500 API Calls instead of just one.
For me, the end-result is what matters: We could not find the project, for whatever reason.

Comment (thisthat) [reply api/pkg/import/handler.go:68]: I agree with your reasoning @christian-kreuzberger-dtx and we should avoid 500 errors here! 
I think we should instead return a 424 error as we have done for the automatic provisioning: https://github.com/keptn/keptn/issues/7149

**Comment by warber** [comment]
@christian-kreuzberger-dtx as I said, no need to put everything into the handler package, you could as well just move the `import` folder beside the handler package. I just don't want to have the `pkg` introduced without a clear purpose.

File: api/pkg/import/project_checker.go
Comment (pchila) [reply api/pkg/import/project_checker.go]: Done

File: api/restapi/configure_keptn.go
Comment (christian-kreuzberger-dtx) [inline api/restapi/configure_keptn.go:65]: That looks like an accidental space :)

File: api/importer/project_checker_test.go
Comment (Raffy23) [inline api/importer/project_checker_test.go:41]: I would prefer to have a function name like `handleRequest` instead of just `f`

File: api/restapi/configure_keptn.go
Comment (pchila) [reply api/restapi/configure_keptn.go]: This is coming from auto-formatting of the code once I touched the file (some comments are not well formatted and somlines are a little long). Undid the format changes for the unaffected parts of the file and pushed a fixup

File: api/test/data/import/sample-package/manifest.yaml
Comment (Raffy23) [inline api/test/data/import/sample-package/manifest.yaml:38]: I think we should use the same parentheses for the templating as in the other files.
```suggestion
      project: "[[ .inputs.project ]]"
```

**Review by christian-kreuzberger-dtx — DISMISSED** [review]
LGTM
@warber @thisthat any objections?

File: api/test/data/import/sample-package/manifest.yaml
Comment (Raffy23) [inline api/test/data/import/sample-package/manifest.yaml:52]: Same here:
```suggestion
      event: "[[ .tasks.create-subscription.context.event ]]"
      subscriptionId: "[[ .tasks.create-subscription.response.id ]]"
```

**Review by Raffy23 — DISMISSED** [review]
LGTM

**Review by thisthat — APPROVED** [review]
LGTM, thank you very much for the contribution! 🦾

**Comment by christian-kreuzberger-dtx** [comment]
The PR is approved, but @pchila please squash and sign the commit. I will re-approve it

**Review by christian-kreuzberger-dtx — DISMISSED** [review]
LGTM

**Review by christian-kreuzberger-dtx — APPROVED** [review]

**Review by christian-kreuzberger-dtx — APPROVED** [review]
LGTM

**Pull request merged**
